File: sdk/go-sdk/middleware.go
package upblit

import (
	"encoding/json"
	"net/http"
	"time"
)

func (s *SDK) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodGet && r.URL.Path == "/health" {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			_ = json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
			return
		}

		traceID := newUUID()
		rootSpanID := newUUID()
		start := now()

		tc := &traceContext{
			TraceID:     traceID,
			CurrentSpan: rootSpanID,
		}

		recorder := &statusRecorder{ResponseWriter: w}
		next.ServeHTTP(recorder, r.WithContext(withTraceContext(r.Context(), tc)))

		s.pushTrace(Trace{
			Timestamp:      now().UTC().Format(time.RFC3339Nano),
			RequestMethod:  "controller:" + r.Method,
			RequestURL:     requestURL(r),
			ResponseStatus: recorder.Status(),
			TraceID:        traceID,
			SpanID:         rootSpanID,
			DurationMs:     now().Sub(start).Milliseconds(),
		})
	})
}

func requestURL(r *http.Request) string {
	if r.URL == nil {
		return ""
	}
	return r.URL.RequestURI()
}

File: sdk/express-sdk/middleware.js
const { randomUUID } = require("crypto");
const { runWithContext } = require("./context");
const transport = require("./service");

function middleware() {
    return (req, res, next) => {
        if (req.method === "GET" && req.path === "/health") {
            return res.status(200).json({ status: "ok" });
        }

        const traceId = randomUUID();
        const rootSpanId = randomUUID();
        const start = Date.now();

        runWithContext({
            traceId,
            currentSpan: rootSpanId
        }, () => {

            res.on("finish", () => {

    

                transport.pushTrace({
                    timestamp: new Date().toISOString(),
                    requestMethod: `controller:${req.method}`,
                    requestURL: req.originalUrl,
                    responseStatus: res.statusCode,
                    traceId,
                    spanId: rootSpanId,
                    parentSpanId: null,
                    durationMs: Date.now() - start
                });
            });

            next();
        });
    };
}

module.exports = middleware;